sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -183,6 +183,118 @@ to throw an exception in such a case just because some other kernel in the same
 device image uses a feature that is not supported by device *D*.
 
 
+### Compiler switch to diagnose optional feature usage
+
+As noted above, the FE compiler does not generally know which kernels will be
+submitted to which devices, and this limits the errors that the FE compiler can
+diagnose.  However, some users may know that all code in their application will
+run on some specific target device, and these users would prefer a compiler
+diagnostic if any kernel uses a feature that is not compatible with that
+device.  This is particularly true for FPGA users who always AOT compile their
+application for a specific FPGA device.
+
+To satisfy this use case, this design adds a new compile-time switch
+`-fsycl-assume-all-kernels-run-on-targets` which is an assertion by the user
+that all device code in the translation unit is expected to be runnable on the
+specified target device(s).
+
+The definition of this switch is:
+
+> `-fsycl-assume-all-kernels-run-on-targets=target,target...`
+>
+> This switch may be passed on the compilation line as an assertion by the
+> user that all device code in the source file (translation unit) is expected
+> to be runnable on the target devices that are listed, and it will not be run
+> on any other target device.  The compiler checks whether the device code uses
+> any features that are incompatible with any of the listed devices.  If so,
+> the compiler raises a diagnostic.
+>
+> This switch works in both ahead-of-time (AOT) and just-in-time (JIT)
+> compilation modes, and it does not automatically enable AOT mode.  In order
+> to perform AOT compilation, pass `-fsycl-targets` on the link line.
+
```

Comment:
Sorry if it's already addressed, but still would like to clarify expected behavior with different combinations:
Note 1: I will use toy target names for simplicity below, 
Note 2: Assuming kernels are compatible with all devices.

1. `-fsycl-assume-all-kernels-run-on-targets=X` is passed, compiler doesn't generate any diagnostic, produces SPIRV.
2.  `-fsycl-assume-all-kernels-run-on-targets=cpu` && `-fsycl-targets=gpu` - compiler issues an error since intersection of targets is zero.
3. `-fsycl-assume-all-kernels-run-on-targets=gpu` && `-fsycl-targets=cpu,gpu` - compiler emits an error here as it cannot produce a binary for `cpu` target?
4. `-fsycl-assume-all-kernels-run-on-targets=cpu,spirv` && `-fsycl-targets=gpu` - error? not sure if `spirv" is a legal target for `-fsycl-assume-all-kernels-run-on-targets`
5.   `-fsycl-assume-all-kernels-run-on-targets=cpu` && `-fsycl-targets=gpu,spirv` - compiler issues no error since spirv is "JIT-covering" cpu target, gpu AOT binary is not generated.

Reply:
> 1. -fsycl-assume-all-kernels-run-on-targets=X - compiler doesn't generate any diagnostic, produces SPIRV.

Correct.

> 2. -fsycl-assume-all-kernels-run-on-targets=cpu && -fsycl-targets=gpu - compiler issues an error since intersection of targets is zero.
> 3. -fsycl-assume-all-kernels-run-on-targets=gpu && -fsycl-targets=cpu,gpu - compiler emits an error here as it cannot produce a binary for cpu target?

I did not expect an error in either of these cases.  We could emit an error, but I think we should be consistent.  Would we also emit an error in these cases, which are similar?

* The application is compiled with `-fsycl-targets=X`, but all kernels in the application use features that are not supported on device X.

* The application is compiled with `-fsycl-targets=X`, but there are no kernels at all in the application.

Currently, I think our expectation is that neither of these cases will result in an error.

> 4. -fsycl-assume-all-kernels-run-on-targets=cpu,spirv && -fsycl-targets=gpu - error? not sure if spirv" is a legal target for -fsycl-assume-all-kernels-run-on-targets`

There is a TODO in the design doc (even from before this PR) about the legal values of "target" in the config file.  Personally, I think we  should not allow generic targets like "spirv" or "gpu".  It's not clear to me what it means to AOT compile for a generic GPU since each GPU has its own native instruction set.  My current thinking is that "spirv" is not allowed as a target for `-fsycl-assume-all-kernels-run-on-targets`, and there is no entry for "spirv" in the config file.

It's not clear to me how / when we resolve this TODO about the entries in the config file.  @kbobrovs, I think you added this TODO.  Will you drive resolution?  Should we have a meeting?

Regardless, I'd prefer not to hold up this PR for this TODO since it existed even before this PR.

> 5. -fsycl-assume-all-kernels-run-on-targets=cpu && -fsycl-targets=gpu,spirv - compiler issues no error since spirv is "JIT-covering" cpu target, gpu AOT binary is not generated.

I agree that this should not cause an error, but I also think that cases 2 and 3 should not cause an error either.

FWIW, I think it was a mistake to allow "spirv" as a target for `-fsycl-targets` because you can't really AOT compile for that target.  Instead, I think the compiler should always generate SPIR-V, regardless of whether you specify `-fsycl-targets`.  This make sense because you need SPIR-V in order to implement some features (like online compilation), and you need SPIR-V to implement specialization constants efficiently.  If there are customers who really don't want SPIR-V and are willing to give up on those features, I'd provide a new switch that disables SPIR-V generation (maybe `-fsycl-no-spirv`).  This is a discussion for another forum, though, since it's not directly related to this PR.

Reply:
> Would we also emit an error in these cases, which are similar?
> 
>     The application is compiled with -fsycl-targets=X, but all kernels in the application use features that are not supported on device X.
> 
>     The application is compiled with -fsycl-targets=X, but there are no kernels at all in the application.

I would expect the first results in error, the second is OK.

> Regardless, I'd prefer not to hold up this PR for this TODO since it existed even before this PR.

I would expect that `-fsycl-targets=X` fails if there are features in the device code that are not supported by target `X`.
So, it sounds that the only thing that  `-fsycl-assume-all-kernels-run-on-targets` adds here is ability to check compatibility when targeting `spirv`(JIT), right?

If so, could we have this semantic?:

1. `-fsycl-check-spirv-compatibility=X` - compiler issues a diagnostic if device code is not compatible with target `X`, produces `spirv`
2.  `-fsycl-check-spirv-compatibility=X && -fsycl-targets=Y` - compiler issues a diagnostic if device code is not compatible with `X` and `Y`, produces `spirv` and `Y`.

Case 2 assumes that we always generate `spirv` target, if we do not we can just ignore `-fsycl-check-spirv-compatibility=X`  notifying a user about that.

Reply:
> So, it sounds that the only thing that -fsycl-assume-all-kernels-run-on-targets adds here is ability to check compatibility when targeting spirv(JIT), right?

No, I don't think this is the case.  Remember that `-fsycl-targets=Y` will silently skip AOT compilation of any kernel that uses a feature not compatible with device `Y`.  The `-fsycl-assume-all-kernels-run-on-targets` switch provides a way for the user to assert that all kernels are supposed to run on target `Y`, and therefore the compiler should issue a diagnostic if any kernel is not compatible with target `Y`.

In addition, the `-fsycl-assume-all-kernels-run-on-targets=X` switch also has the semantic that the kernels in the TU will *only* be submitted to device `X`.  If the application attempts to submit the kernel to some other device `Z`, the runtime raises an exception.  This is the case even if the kernel happens to be compatible with device `Z`.

Reply:
>>  Remember that -fsycl-targets=Y will silently skip AOT compilation of any kernel that uses a feature not compatible with device Y

Didn't know that, this behavior doesn't sound user friendly as, I believe, if a user specifies `-fsycl-targets=Y` they expect they their app can run on target "Y". Can we change it?

>> In addition, the -fsycl-assume-all-kernels-run-on-targets=X switch also has the semantic that the kernels in the TU will only be submitted to device X. If the application attempts to submit the kernel to some other device Z, the runtime raises an exception. This is the case even if the kernel happens to be compatible with device Z.

Could you please clarify how it can be useful?


BTW.
Do not think my comments should block this PR.

Reply:
> Didn't know that, this behavior doesn't sound user friendly as, I believe, if a user specifies -fsycl-targets=Y they expect they their app can run on target "Y". Can we change it?

I agree that users expect that their application can run on target "Y", but this is not the same as saying that *every* kernel in the application can  run on target "Y".  One of the goals of SYCL is to encourage heterogenous computing, where the application takes advantage of the best device for each algorithm.  We want applications to be able to run some kernels on device "X" and run other kernels on device "Y", so an application can have a mixture of kernels that are coded for different devices.

If the user specifies `-fsycl-targets=Y`, I think this should simply pre-compile kernels for that device in order to avoid the JIT compilation overhead.  In the case where the application has a mixture of kernels for both device "X" and "Y", the user might specify `-fsycl-targets=X,Y`.  This would pre-compile kernels for those two devices.  It will be very normal, though, for the application to contain some kernels that use features specific to device "X" (or for device "Y").  The AOT compiler can't raise an error when this happens, or AOT will be useless for applications that contain a mixture of kernels for different devices.  Therefore, the behavior of `-fsycl-targets` should be to silently skip AOT compilation of kernels that are incompatible with the specified AOT device.

There is more information about this topic in the requirements section at the top of this design doc.  See the section "Diagnostics from the FE compiler", and the paragraph that starts "Note that this behavior does not change when the compiler runs in AOT mode".

> Could you please clarify how it can be useful?

When the user specifies `-fsycl-assume-all-kernels-run-on-targets=X` they are asserting that all kernels in the TU are expected to run on device "X" and that the application promises to never run the kernels on any other device.  These two assertions together allow the compiler to safely bundle all the kernels together into a single device image.  Bundling all the kernels together like this, though, is unsafe without these assertions (doing so could lead to confusing errors at runtime).  In order to avoid these confusing errors, the runtime checks that the application obeys the assertion: If the TU is compiled with `-fsycl-assume-all-kernels-run-on-targets=X` and a kernel in the TU is run on a device other than "X", the runtime raises an exception.

Reply:
> It's not clear to me how / when we resolve this TODO about the entries in the config file. @kbobrovs, I think you added this TODO. Will you drive resolution? Should we have a meeting?

@gmlueck - sorry, missed this one. Yes, I will drive this, we need to finalize on target identification.